File: android/src/main/java/app/formstr/plugins/securekeystorage/SecureKeyStoragePlugin.java
package app.formstr.plugins.securekeystorage;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.security.crypto.EncryptedSharedPreferences;
import androidx.security.crypto.MasterKey;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "SecureKeyStorage")
public class SecureKeyStoragePlugin extends Plugin {
    private static final String PREFS_NAME = "secure_key_storage";

    private SharedPreferences getSecurePreferences() throws Exception {
        Context context = getContext();
        MasterKey masterKey = new MasterKey.Builder(context)
                .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                .build();

        return EncryptedSharedPreferences.create(
                context,
                PREFS_NAME,
                masterKey,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        );
    }

    @PluginMethod
    public void set(PluginCall call) {
        String key = call.getString("key");
        String value = call.getString("value");

        if (key == null || key.isEmpty()) {
            call.reject("Key is required");
            return;
        }

        if (value == null) {
            call.reject("Value is required");
            return;
        }

        try {
            getSecurePreferences().edit().putString(key, value).apply();
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to securely store value", e);
        }
    }

    @PluginMethod
    public void get(PluginCall call) {
        String key = call.getString("key");

        if (key == null || key.isEmpty()) {
            call.reject("Key is required");
            return;
        }

        try {
            String value = getSecurePreferences().getString(key, null);
            JSObject result = new JSObject();
            result.put("value", value);
            call.resolve(result);
        } catch (Exception e) {
            call.reject("Failed to securely load value", e);
        }
    }

    @PluginMethod
    public void remove(PluginCall call) {
        String key = call.getString("key");

        if (key == null || key.isEmpty()) {
            call.reject("Key is required");
            return;
        }

        try {
            getSecurePreferences().edit().remove(key).apply();
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to securely remove value", e);
        }
    }
}

File: ios/Sources/SecureKeyStoragePlugin/SecureKeyStoragePlugin.swift
import Foundation
import Capacitor

@objc(SecureKeyStoragePlugin)
public class SecureKeyStoragePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "SecureKeyStoragePlugin"
    public let jsName = "SecureKeyStorage"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "set", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "get", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "remove", returnType: CAPPluginReturnPromise)
    ]

    private let keychain = KeychainStorage()

    @objc func set(_ call: CAPPluginCall) {
        guard let key = call.getString("key"), !key.isEmpty else {
            call.reject("Key is required")
            return
        }

        guard let value = call.getString("value") else {
            call.reject("Value is required")
            return
        }

        do {
            try keychain.set(value, for: key)
            call.resolve()
        } catch {
            call.reject("Failed to securely store value", nil, error)
        }
    }

    @objc func get(_ call: CAPPluginCall) {
        guard let key = call.getString("key"), !key.isEmpty else {
            call.reject("Key is required")
            return
        }

        do {
            let value = try keychain.get(key)
            call.resolve([
                "value": value as Any
            ])
        } catch {
            call.reject("Failed to securely load value", nil, error)
        }
    }

    @objc func remove(_ call: CAPPluginCall) {
        guard let key = call.getString("key"), !key.isEmpty else {
            call.reject("Key is required")
            return
        }

        do {
            try keychain.remove(key)
            call.resolve()
        } catch {
            call.reject("Failed to securely remove value", nil, error)
        }
    }
}
